engine/src/main/java/org/hibernate/validator/spi/nodenameprovider/PropertyNodeNameProvider.java

```diff
@@ -0,0 +1,11 @@
+/*
+ * Hibernate Validator, declare and validate application constraints
+ *
+ * License: Apache License, Version 2.0
+ * See the license.txt file in the root directory or <http://www.apache.org/licenses/LICENSE-2.0>.
+ */
+package org.hibernate.validator.spi.nodenameprovider;
+
+public interface PropertyNodeNameProvider {
+	String getName(String propertyName, Object object);
```

Comment:
hmm, that's an interesting choice of a contract. It's a nice way to not rely neither on reflection nor on our current `internal.properties.*` wrappers. I think though there might be a problem with JSON/Map validation case. An implementation of a `PropertyNodeNameProvider` would need to check the type of an object passed and determine how to proceed based on that. And as we plan to support any kind of property holders the implementations of `PropertyNodeNameProvider` would need to know about all of them.
Using `internal.properties.*` types would probably solve the problem but as @gsmet mentioned, these types are not yet ready to be public, and they might not become such.

Reply:
The idea we have is that we provide the contract, an extension point, for property name resolution, and let others implement it (not have the implementations part of the `hibernate-validator`). That way, it's up to the one who is implementing the contract to decide how to resolve the name (depending on the use case). One example of implementation is to use reflection and resolve the name from the annotation.

Could you please describe in more details your concern regarding JSON/Map validation case?

Reply:
Yes sure. Let me try to show it with the example. Let's assume we have next bean:
```java
class SomeBean {
    @Valid 
    private SomeOtherBean someOtherBean;

    @NotNull
    private String string;

    @ValidPropertyHolder(...)
    private Map<String, ?> mapAsPropertyHolder;

    @ValidPropertyHolder(...)
    private JsonObject someJsonAsPropertyHolder;

    // .... anything else
}
```
fields annotated with `ValidPropertyHolder` are considered to be "property holders" and each their property could have own set of validation constraint rules. Hence `PropertyNodeNameProvider` would also be called on properties of these structures. By properties I mean something like:
```java
mapAsPropertyHolder.get("SomePropertyKey");
```
 And what I was concerned about is that in case of the current `PropertyNodeNameProvider` interface that you've provided, you could get in situation where object is either `Map` or some `JsonObject` and an implementation would need to check for all these possible cases before making any actions. 
Going to the example that you have in the tests, in case of the bean above, you would first need to check if passed object is a `Map`, if so - do some custom logic for it (or simply return the key) then check if maybe it's a JSON, and in the end if it doesn't fall under any "special" cases you would try to get a field or a getter matching the given property name. 
While it's all doable, the implementation of such `PropertyNodeNameProvider` could grow big and would be dependent on all these cases. Using something like `internal.properties.*`, in theory, should prevent this from happening as it all we be abstracted out.

Reply:
Thank you for the example.

I have updated the PR, and added a new comment addressing your concerns: https://github.com/hibernate/hibernate-validator/pull/1012#discussion_r250205593